File: admin/assets/examples/licora-v2/flutter/licora_v2_client.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

import 'package:cryptography/cryptography.dart';
import 'package:http/http.dart' as http;

/// Licora Secure API v2 lifecycle reference for Flutter/Dart.
///
/// This creates an ephemeral P-256 device for developer testing and deactivates
/// it before exit. A production app must persist the private key and rotated
/// refresh token in platform secure storage, and verify LICORA-V2/RS256 tokens
/// with the pinned Licora server public key before trusting token claims.
class LicoraV2Client {
  LicoraV2Client(this.baseUrl, this.appId, this.appVersion);

  final String baseUrl;
  final String appId;
  final String appVersion;
  final Ecdsa _ecdsa = Ecdsa.p256(Sha256());
  final Random _random = Random.secure();
  SimpleKeyPair? _keyPair;
  String? _publicPem;
  late final String deviceId = 'flutter-${_randomBytes(16).map((b) => b.toRadixString(16).padLeft(2, '0')).join()}';

  List<int> _randomBytes(int count) => List<int>.generate(count, (_) => _random.nextInt(256));
  String _endpoint(String name) => '${baseUrl.replaceFirst(RegExp(r'/+$'), '')}/api/v2/$name.php';
  String _b64url(List<int> data) => base64Url.encode(data).replaceAll('=', '');

  Future<String> _sha256Hex(List<int> data) async {
    final hash = await Sha256().hash(data);
    return hash.bytes.map((b) => b.toRadixString(16).padLeft(2, '0')).join();
  }

  Future<void> _ensureKey() async {
    if (_keyPair != null) return;
    _keyPair = await _ecdsa.newKeyPair();
    final publicKey = await _keyPair!.extractPublicKey();
    final raw = publicKey.bytes;
    if (raw.length != 65 || raw.first != 0x04) {
      throw StateError('Unexpected P-256 public-key encoding; expected uncompressed 65-byte point.');
    }
    // X.509 SubjectPublicKeyInfo prefix for id-ecPublicKey + prime256v1.
    const prefixHex = '3059301306072a8648ce3d020106082a8648ce3d030107034200';
    final prefix = <int>[];
    for (var i = 0; i < prefixHex.length; i += 2) {
      prefix.add(int.parse(prefixHex.substring(i, i + 2), radix: 16));
    }
    final der = Uint8List.fromList([...prefix, ...raw]);
    final encoded = base64.encode(der);
    final lines = <String>[];
    for (var i = 0; i < encoded.length; i += 64) {
      lines.add(encoded.substring(i, min(i + 64, encoded.length)));
    }
    _publicPem = '-----BEGIN PUBLIC KEY-----\n${lines.join('\n')}\n-----END PUBLIC KEY-----\n';
  }

  List<int> _ecdsaDer(List<int> signature) {
    if (signature.isNotEmpty && signature.first == 0x30) return signature;
    if (signature.length != 64) throw StateError('Unexpected P-256 signature length: ${signature.length}');
    List<int> integer(List<int> value) {
      var i = 0;
      while (i < value.length - 1 && value[i] == 0) i++;
      var bytes = value.sublist(i);
      if ((bytes.first & 0x80) != 0) bytes = [0, ...bytes];
      return [0x02, bytes.length, ...bytes];
    }
    final r = integer(signature.sublist(0, 32));
    final s = integer(signature.sublist(32, 64));
    final body = [...r, ...s];
    return [0x30, body.length, ...body];
  }

  String _jti(String token) {
    final parts = token.split('.');
    if (parts.length != 3) throw StateError('Licora returned a malformed access token.');
    final payload = jsonDecode(utf8.decode(base64Url.decode(base64Url.normalize(parts[1])))) as Map<String, dynamic>;
    return payload['jti'] as String;
  }

  Future<Map<String, dynamic>> _post(String name, Map<String, dynamic> payload, String context, [String? accessToken]) async {
    await _ensureKey();
    final url = _endpoint(name);
    final bodyText = jsonEncode(payload);
    final body = utf8.encode(bodyText);
    final timestamp = DateTime.now().toUtc().millisecondsSinceEpoch ~/ 1000;
    final nonce = _b64url(_randomBytes(18));
    final canonical = ['POST', Uri.parse(url).path, '$timestamp', nonce, await _sha256Hex(body), context].join('\n');
    final signature = await _ecdsa.sign(utf8.encode(canonical), keyPair: _keyPair!);
    final response = await http.post(Uri.parse(url), body: body, headers: {
      'Content-Type': 'application/json',
      'X-Licora-Timestamp': '$timestamp',
      'X-Licora-Nonce': nonce,
      'X-Licora-Device-Signature': _b64url(_ecdsaDer(signature.bytes)),
      if (accessToken != null) 'Authorization': 'Bearer $accessToken',
    });
    final data = jsonDecode(response.body) as Map<String, dynamic>;
    if (data['success'] != true) throw StateError('Licora error ${data['code'] ?? 'UNKNOWN'} (HTTP ${response.statusCode})');
    return data;
  }

  Future<Map<String, dynamic>> activate(String licenseKey) async {
    await _ensureKey();
    return _post('activate', {
      'license_key': licenseKey, 'app_id': appId, 'app_version': appVersion,
      'device_id': deviceId, 'device_public_key': _publicPem,
    }, 'activate:$appId');
  }

  Future<Map<String, dynamic>> status(String accessToken) => _post('status', {}, _jti(accessToken), accessToken);
  Future<Map<String, dynamic>> refresh(String refreshToken) async => _post('refresh', {
    'refresh_token': refreshToken, 'app_version': appVersion,
  }, 'refresh:${await _sha256Hex(utf8.encode(refreshToken))}');
  Future<Map<String, dynamic>> deactivate(String accessToken) => _post('deactivate', {}, _jti(accessToken), accessToken);

  static Future<void> lifecycleTest({required String baseUrl, required String appId, required String licenseKey, String appVersion = '1.0.0'}) async {
    final client = LicoraV2Client(baseUrl, appId, appVersion);
    String accessToken = '';
    try {
      final activated = await client.activate(licenseKey); accessToken = activated['access_token'] as String; var refreshToken = activated['refresh_token'] as String; print('[PASS] activate');
      await client.status(accessToken); print('[PASS] status');
      final refreshed = await client.refresh(refreshToken); accessToken = refreshed['access_token'] as String; refreshToken = refreshed['refresh_token'] as String; print('[PASS] refresh (rotated)');
      await client.status(accessToken); print('[PASS] status-after-refresh');
      await client.deactivate(accessToken); accessToken = ''; print('[PASS] deactivate');
    } finally {
      if (accessToken.isNotEmpty) { try { await client.deactivate(accessToken); print('[INFO] cleanup deactivate completed'); } catch (_) { print('[WARN] cleanup deactivate failed'); } }
    }
  }
}

File: admin/assets/examples/licora-v2/react-native/licoraV2Client.js
import crypto from 'react-native-quick-crypto';
import { Buffer } from 'buffer';

/**
 * Licora Secure API v2 lifecycle reference for React Native.
 *
 * This developer/test helper creates an ephemeral P-256 credential and
 * deactivates it. Production apps must store the private key and rotated
 * refresh token with platform-secure storage and verify LICORA-V2/RS256
 * access-token signatures using the pinned Licora server public key.
 */

const b64url = (data) => Buffer.from(data).toString('base64').replace(/=/g, '').replace(/\+/g, '-').replace(/\//g, '_');
const sha256 = (data) => crypto.createHash('sha256').update(data).digest('hex');
const randomHex = (bytes) => crypto.randomBytes(bytes).toString('hex');

function jwtJti(token) {
  const parts = token.split('.');
  if (parts.length !== 3) throw new Error('Licora returned a malformed access token.');
  const payload = JSON.parse(Buffer.from(parts[1], 'base64url').toString('utf8'));
  return String(payload.jti);
}

export class LicoraV2Client {
  constructor(baseUrl, appId, appVersion = '1.0.0') {
    this.baseUrl = baseUrl.replace(/\/+$/, '');
    this.appId = appId;
    this.appVersion = appVersion;
    const pair = crypto.generateKeyPairSync('ec', {
      namedCurve: 'prime256v1',
      publicKeyEncoding: { type: 'spki', format: 'pem' },
      privateKeyEncoding: { type: 'pkcs8', format: 'pem' },
    });
    this.publicKey = pair.publicKey;
    this.privateKey = pair.privateKey;
    this.deviceId = `rn-${randomHex(16)}`;
  }

  endpoint(name) { return `${this.baseUrl}/api/v2/${name}.php`; }

  headers(url, body, context, accessToken) {
    const timestamp = Math.floor(Date.now() / 1000);
    const nonce = b64url(crypto.randomBytes(18));
    const canonical = ['POST', new URL(url).pathname || '/', timestamp, nonce, sha256(body), context].join('\n');
    const signer = crypto.createSign('SHA256');
    signer.update(canonical, 'utf8'); signer.end();
    const signature = signer.sign({ key: this.privateKey, dsaEncoding: 'der' });
    return {
      'Content-Type': 'application/json',
      'X-Licora-Timestamp': String(timestamp),
      'X-Licora-Nonce': nonce,
      'X-Licora-Device-Signature': b64url(signature),
      ...(accessToken ? { Authorization: `Bearer ${accessToken}` } : {}),
    };
  }

  async post(name, payload, context, accessToken = '') {
    const url = this.endpoint(name);
    const body = JSON.stringify(payload);
    const response = await fetch(url, { method: 'POST', headers: this.headers(url, Buffer.from(body, 'utf8'), context, accessToken), body });
    const data = await response.json();
    if (!data.success) throw new Error(`Licora error ${data.code || 'UNKNOWN'} (HTTP ${response.status})`);
    return data;
  }

  activate(licenseKey) { return this.post('activate', { license_key: licenseKey, app_id: this.appId, app_version: this.appVersion, device_id: this.deviceId, device_public_key: this.publicKey }, `activate:${this.appId}`); }
  status(accessToken) { return this.post('status', {}, jwtJti(accessToken), accessToken); }
  refresh(refreshToken) { return this.post('refresh', { refresh_token: refreshToken, app_version: this.appVersion }, `refresh:${sha256(Buffer.from(refreshToken, 'utf8'))}`); }
  deactivate(accessToken) { return this.post('deactivate', {}, jwtJti(accessToken), accessToken); }
}

export async function lifecycleTest({ baseUrl, appId, licenseKey, appVersion = '1.0.0' }) {
  const client = new LicoraV2Client(baseUrl, appId, appVersion);
  let accessToken = '';
  try {
    const activated = await client.activate(licenseKey); accessToken = activated.access_token; let refreshToken = activated.refresh_token; console.log('[PASS] activate');
    await client.status(accessToken); console.log('[PASS] status');
    const refreshed = await client.refresh(refreshToken); accessToken = refreshed.access_token; refreshToken = refreshed.refresh_token; console.log('[PASS] refresh (rotated)');
    await client.status(accessToken); console.log('[PASS] status-after-refresh');
    await client.deactivate(accessToken); accessToken = ''; console.log('[PASS] deactivate');
  } finally {
    if (accessToken) { try { await client.deactivate(accessToken); console.log('[INFO] cleanup deactivate completed'); } catch (_) { console.warn('[WARN] cleanup deactivate failed'); } }
  }
}
